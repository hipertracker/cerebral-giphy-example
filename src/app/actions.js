import _map from 'lodash/map'

// it would no be necessary for GraphQL end-point
function normalizeData (response) {
  const data = _map(response.data, item => {
    return {
      id: item.id,
      title: item.title,
      gif_url: item.images.fixed_width_small_still.gif_url,
      width: parseInt(item.images.fixed_width_small_still.width),
      height: parseInt(item.images.fixed_width_small_still.height),
    }
  })
  return {...response, data}
}

export function fetchData ({giphy, path, state}) {
  const page = state.get('pagination.page')
  const limit = state.get('fetchLimit')
  const params = {
    q: state.get('query'),
    offset: (page - 1) * limit,
    limit,
    sort: 'recent',
  }
  return giphy.search(params).
    then(response => {
      const data = normalizeData(response)
      return path.ok(data)
    }).catch(error => path.failed({error}))
}

export function updateState ({props, state}) {
  // data piped and destructured from loadData()

  const {
    data,
    pagination: {
      count,
      total_count: totalCount,
    },
  } = props

  if (!data) return

  const totalPages = Math.floor(totalCount / state.get('fetchLimit'))

  // normalize array to object for VDOM optimization
  const groupById = data.reduce(
    (acc, item) => ({...acc, [item.id]: item}),
    {},
  )

  // update UI state tree
  state.set('data', groupById)
  state.set('pagination.count', count)
  state.set('pagination.totalCount', totalCount)

  // API limit, max 100 pages
  state.set('pagination.totalPages', totalPages > 100 ? 100 : totalPages)

  // set the first page if needed
  const page = state.get('pagination.page')
  if (!page && count > 0) {
    state.set('pagination.page', 1)
  }
}


