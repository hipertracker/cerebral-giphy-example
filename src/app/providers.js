import { Provider } from 'cerebral'
import GiphyApiClient from 'giphy-js-sdk-core'

const apiKey = process.env.REACT_APP_GIPHY_API
const client = GiphyApiClient(apiKey)

export const Giphy = Provider({
  search (...params) {
    return client.search('gifs',...params)
  },
})



