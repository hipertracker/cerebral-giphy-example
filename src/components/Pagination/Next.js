import { connect } from '@cerebral/react'
import { signal, state } from 'cerebral/tags'
import PropTypes from 'prop-types'
import React from 'react'

export default connect(
  {
    page: state`pagination.page`,
    totalPages: state`pagination.totalPages`,
    pageChanged: signal`pageChanged`,
  },
  class Next extends React.PureComponent {
    static propTypes = {
      page: PropTypes.number.isRequired,
      totalPages: PropTypes.number.isRequired,
      pageChanged: PropTypes.func.isRequired,
    }

    clicked = () => {
      const value = this.props.page + 1
      this.props.pageChanged({value})
    }

    render () {
      const {page, totalPages} = this.props
      if (page === totalPages) {
        return (
          <a className="pagination-next" disabled>
            Next
          </a>
        )
      }
      return (
        <a className="pagination-next" onClick={this.clicked}>
          Next
        </a>)
    }
  },
)
