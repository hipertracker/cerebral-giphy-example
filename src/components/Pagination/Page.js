import { connect } from '@cerebral/react'
import { signal } from 'cerebral/tags'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

export default connect(
  {
    pageChanged: signal`pageChanged`,
  },
  class Page extends React.PureComponent {
    static propTypes = {
      pageChanged: PropTypes.func.isRequired,
    }

    clicked = () => {
      this.props.pageChanged({value: this.props.value})
    }

    render () {
      const {value, current} = this.props
      const css = classNames('pagination-link', {
        'is-current': value === current,
      })
      return (
        <li>
          <a className={css} aria-label={`Goto page ${value}`}
             onClick={this.clicked}>
            {value}
          </a>
        </li>
      )
    }
  },
)


