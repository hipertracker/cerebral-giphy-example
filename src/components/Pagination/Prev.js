import { connect } from '@cerebral/react'
import { signal, state } from 'cerebral/tags'
import PropTypes from 'prop-types'
import React from 'react'

export default connect(
  {
    page: state`pagination.page`,
    pageChanged: signal`pageChanged`,
  },
  class Prev extends React.PureComponent {
    static propTypes = {
      page: PropTypes.number.isRequired,
      pageChanged: PropTypes.func.isRequired,
    }

    clicked = () => {
      const value = this.props.page - 1
      this.props.pageChanged({value})
    }

    render () {
      if (this.props.page === 1) {
        return (
          <a className="pagination-previous" disabled>
            Previous
          </a>
        )
      }

      return (
        <a className="pagination-previous" onClick={this.clicked}>
          Previous
        </a>)
    }
  },
)
