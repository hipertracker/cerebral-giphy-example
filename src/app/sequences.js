// see: https://cerebraljs.com/docs/introduction/signals.html

import { set, when } from 'cerebral/operators'
import { props, state } from 'cerebral/tags'
import * as actions from './actions'

const loadData = [
  actions.fetchData, //async
  {
    ok: actions.updateState,
    failed: set(state`errors`, props`error`),
  },
]

export const routeToHome = loadData

export const recoverNetworkFailure = set(state`errors`, props`response`)

export const queryChanged = [
  set(state`pagination.page`, 1),
  set(state`query`, props`value`),
  ...loadData,
]

export const pageChanged = [
  set(state`pagination.page`, props`value`),
  ...loadData,
]

export const toggleBigPicture = [
  set(state`pictureId`, props`value`),
  when(state`data`, data => !!data), {
    true: [],
    false: loadData,
  },
]

export const closeModal = set(state`pictureId`, null)

export const clearErrors = set(state`errors`, null)
