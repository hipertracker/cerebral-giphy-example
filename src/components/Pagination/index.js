import { connect } from '@cerebral/react'
import { signal, state } from 'cerebral/tags'
import PropTypes from 'prop-types'
import React from 'react'
import Next from './Next'
import Page from './Page'
import Prev from './Prev'

export default connect(
  {
    page: state`pagination.page`,
    totalPages: state`pagination.totalPages`,
    pageChanged: signal`pageChanged`,
  },
  class Pagination extends React.PureComponent {
    static propTypes = {
      page: PropTypes.number.isRequired,
      totalPages: PropTypes.number.isRequired,
    }

    render () {
      const {page, totalPages} = this.props

      if (!totalPages) return null

      let nr = [page - 1, page, page + 1]
      if (page === totalPages) {
        nr = [page - 3, page - 2, page - 1]
      } else if (page < 3) {
        nr = [2, 3, 4]
      } else if (page === totalPages - 1) {
        nr = [page - 2, page - 1, page]
      }
      return (
        <nav className="pagination is-centered" aria-label="pagination">
          <Prev/>
          <Next/>
          <ul className="pagination-list">
            <Page value={1} current={page}/>
            <li><span className="pagination-ellipsis">&hellip;</span></li>
            <Page value={nr[0]} current={page}/>
            <Page value={nr[1]} current={page}/>
            <Page value={nr[2]} current={page}/>
            <li><span className="pagination-ellipsis">&hellip;</span></li>
            <Page value={totalPages} current={page}/>
          </ul>
        </nav>
      )
    }
  },
)
