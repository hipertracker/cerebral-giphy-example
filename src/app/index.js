import * as providers  from './providers'
import * as sequences from './sequences'

import HttpProvider, { HttpProviderError } from '@cerebral/http'

import { Module } from 'cerebral'
import { redirectToSignal } from '@cerebral/router/operators'
import router from './router'

export default Module({
  state: {
    fetchLimit: 25,
    query: 'cats',
    data: null,
    pagination: {
      page: 1,
      totalCount: 0,
      totalPages: 0,
    },
    pictureId: null,
    pictureZoom: 40,
    errors: null,
  },
  signals: {
    homeRouted: sequences.routeToHome,
    networkFailed: sequences.recoverNetworkFailure,
    queryChanged: sequences.queryChanged,
    pageChanged: sequences.pageChanged,
    pictureClicked: sequences.toggleBigPicture,
    modalClosed: sequences.closeModal,
    errorMessageClosed: sequences.clearErrors,
  },
  modules: {
    router,
  },
  providers: {
    // see: https://cerebraljs.com/docs/introduction/debugger.html
    http: HttpProvider({
      baseUrl: process.env.NODE_ENV === 'development'
        ? 'http://localhost:3000' : '',
      headers: {
        'Content-Type': 'application/json; charset=UTF-8',
        'Accept': 'application/json',
      },
      withCredentials: false,
    }),
    giphy: providers.Giphy,
  },
  catch: [
    [HttpProviderError, redirectToSignal('networkFailed')],
  ],
})
