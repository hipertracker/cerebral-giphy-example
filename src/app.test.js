import { Container } from '@cerebral/react'
import { Snapshot } from 'cerebral/test'
import React from 'react'
import renderer from 'react-test-renderer'
import app from './app'
import Pagination from './components/Pagination'

// snapshot testing is in beta
test('homeRouted', () => {
  return Snapshot(app).
    mutate('set', 'query', 'cats').
    mutate('set', 'pagination.page', 2).
    mutate('set', 'pictureId', null).
    mockResolvedPromise('giphy.search', {
      data: [],
      meta: {msg: 'OK', response_id: '5a71c9507351665a778766a7', status: 200},
      pagination: {count: 25, offset: 25, total_count: 7576},
    }).
    run('homeRouted', {q: 'cats', page: 2, id: null}).
    then(snapshot => {
      const tree = renderer.create(
        <Container controller={snapshot.controller}>
          <Pagination/>
        </Container>,
      ).toJSON()
      expect(tree).toMatchSnapshot()
      expect(snapshot.get()).toMatchSnapshot()
    })
})


