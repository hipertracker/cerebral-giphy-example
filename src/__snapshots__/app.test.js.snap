// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`homeRouted 1`] = `
<nav
  aria-label="pagination"
  className="pagination is-centered"
>
  <a
    className="pagination-previous"
    onClick={[Function]}
  >
    Previous
  </a>
  <a
    className="pagination-next"
    onClick={[Function]}
  >
    Next
  </a>
  <ul
    className="pagination-list"
  >
    <li>
      <a
        aria-label="Goto page 1"
        className="pagination-link"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-label="Goto page 2"
        className="pagination-link is-current"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li>
      <a
        aria-label="Goto page 3"
        className="pagination-link"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li>
      <a
        aria-label="Goto page 4"
        className="pagination-link"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-label="Goto page 100"
        className="pagination-link"
        onClick={[Function]}
      >
        100
      </a>
    </li>
  </ul>
</nav>
`;

exports[`homeRouted 2`] = `
Array [
  Action {
    "name": "fetchData",
  },
  Provider {
    "args": Array [
      Object {
        "limit": 25,
        "offset": 25,
        "q": "cats",
        "sort": "recent",
      },
    ],
    "method": "giphy.search",
  },
  Path {
    "name": "ok",
  },
  Action {
    "name": "updateState",
  },
  Mutation {
    "args": Array [
      Object {},
    ],
    "method": "set",
    "path": "data",
  },
  Mutation {
    "args": Array [
      25,
    ],
    "method": "set",
    "path": "pagination.count",
  },
  Mutation {
    "args": Array [
      7576,
    ],
    "method": "set",
    "path": "pagination.totalCount",
  },
  Mutation {
    "args": Array [
      100,
    ],
    "method": "set",
    "path": "pagination.totalPages",
  },
]
`;
